### PR TITLE
Add trace ray ground Lua functions

### PR DESCRIPTION
**Pull request by sprunk**
Extracted from #1624.
Depends on #1705.

### CODE REVIEW

**Review by loveridge — APPROVED** [review]

**Comment by saurtron** [comment]
Hey, created a test widget for this: [dbg_gui_test_rays.lua](https://github.com/saurtron/Beyond-All-Reason/blob/test-lua-rays/luaui/Widgets/dbg_gui_test_rays.lua).

1. Put in your Widgets dir
2. Start skirmish
3. Enable widget "TestRays"
4. Click in different places
   - the widget will Echo what was traced by the ray (units, features or ground in that order, if units gets found then it doesnt go on, and so on)

**Comment by saurtron** [comment]
oops added some comments before that were actually for https://github.com/beyond-all-reason/spring/pull/1624

sorry, so many traceray PRs XD

will test this one soon too

File: rts/Lua/LuaSyncedRead.cpp [hunk 1]
```diff
@@ -8592,6 +8596,80 @@ int LuaSyncedRead::GetUnitScriptNames(lua_State* L)
...
+ * @treturn number rayLength
+ * @treturn number posX
+ * @treturn number posY
+ * @treturn number posZ
+ */
+int LuaSyncedRead::TraceRayGroundInDirection(lua_State* L)
+{
+	const float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	const auto dir = float3(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6)).Normalize();
+	const float maxLen = luaL_optfloat(L, 7, 999999.f);
```
Comment (saurtron) [inline rts/Lua/LuaSyncedRead.cpp:8637]: maybe default to FarPlane*1.4 as most of the code is doing, but not really an issue.

**Review by saurtron — COMMENTED** [review]
Added a small nipick, otherwise I found this to work fine.

Created the following widget to test: [dbg_gui_test_ground_rays.lua](https://github.com/saurtron/Beyond-All-Reason/blob/test-lua-ground-rays/luaui/Widgets/dbg_gui_test_ground_rays.lua)

- Start skirmish
- Enable widget "TestRays Ground"
- Click somewhere, you will get:
  - echo of ground intersection
  - echo of water intersection
  - also calculate ground intersection with raybetween to see it gives the same result
  - do ray between first two selected units and send the first to attack ground the found position with some offsets in y so its likely to find the middle point
    - if not nselected == 2, then it will just get the first unit from AllUnits, and do a ray intersection to the clicked point, offseting ground.y +10, and unit.y -10, send it there to attack ground.
- tells you the water depth inferred from both water and ground rays

**Review by saurtron — APPROVED** [review]

File: rts/Lua/LuaSyncedRead.cpp (hunk 1 above)
Comment (sprunk) [reply rts/Lua/LuaSyncedRead.cpp:8637]: I think defaulting to camera's FarPlane*1.4 is done for traces that have the camera as the origin, but here you can have an arbitrary point as the origin.

**Pull request merged**